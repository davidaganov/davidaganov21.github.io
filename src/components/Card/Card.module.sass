.project
  display: flex
  flex-direction: column
  height: 23rem
  border: .1rem solid var(--color-accent)
  box-shadow: .1rem .1rem 0 .1rem var(--color-accent)
  padding: 2.5rem 2rem
  margin: 1rem
  transition: all .2s
  &:hover
    transform: translateY(-.2rem)
    box-shadow: .1rem .3rem 0 .1rem var(--color-accent)

.title
  text-transform: capitalize
  font: bold 2rem var(--main-font)
  color: var(--color-accent)

.description
  margin: 1rem 0 2rem
  color: var(--color-white-700)
  word-wrap: anywhere
  display: -webkit-box
  -webkit-line-clamp: 2
  -webkit-box-orient: vertical
  overflow: hidden

.bottom
  margin-top: auto

.link
  margin-left: -.2rem
  &--live
    position: relative
    margin-left: 1.8rem
    &::before
      position: absolute
      content: ""
      width: .6rem
      height: .6rem
      left: -1.3rem
      top: .1rem
      bottom: 0
      margin: auto 0
      border-radius: 1rem
      background-color: var(--color-white-700)

.tags
  display: flex
  flex-wrap: wrap
  margin-top: .5rem
  padding-top: .5rem
  border-top: .1rem solid var(--color-white-700)
  li
    margin: 1rem 1rem 0 0
    padding: 0 .6rem .2rem .5rem
    text-transform: uppercase
    font: normal 1.6rem var(--second-font)
    border: .1rem solid var(--color-white-700)
    color: var(--color-white-700)
    transition: all .2s
