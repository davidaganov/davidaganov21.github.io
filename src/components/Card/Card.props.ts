import { DetailedHTMLProps, HTMLAttributes } from "react"
import { RepoProps } from "../../interfaces"

export interface CardProps extends DetailedHTMLProps<HTMLAttributes<HTMLLIElement>, HTMLLIElement> {
  card: RepoProps
}
