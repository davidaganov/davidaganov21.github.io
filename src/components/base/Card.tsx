import { CardProps } from "../../interfaces"

export default function Card({ name, description, html_url, topics }: CardProps) {
  const renderTags = (tags: string[]) => {
    const items = tags.map((tag, i) => {
      return <li key={i}>{tag}</li>
    })

    return <ul className="project__tags">{items}</ul>
  }

  const tags = renderTags(topics)

  return (
    <li className="work__project project">
      <h3 className="project__title">{name.replace(/[.\-/\\\s]/g, " ")}</h3>
      <p className="project__description">{description}</p>
      <div className="project__bottom">
        <a
          href={html_url}
          className="project__link inline-link"
          target="_blank"
          rel="noreferrer"
        >
          Go to repository
        </a>
        {tags}
      </div>
    </li>
  )
}
