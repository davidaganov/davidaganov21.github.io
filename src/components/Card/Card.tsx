import { CardProps } from "./Card.props"
import styles from "./Card.module.sass"
import cn from "classnames"

export const Card = ({ card, className }: CardProps): JSX.Element => {
  const { name, description, html_url, topics, homepage } = card

  const renderTags = () => {
    const items = topics.map((tag, i) => {
      return <li key={i}>{tag}</li>
    })

    return <ul className={styles.tags}>{items}</ul>
  }

  const renderDemo = () => {
    if (homepage) {
      return (
        <a
          href={homepage}
          className={cn(styles.link, styles["link--live"], "inline-link")}
          target="_blank"
          rel="noreferrer"
        >
          Live demo
        </a>
      )
    }
  }

  const tags = renderTags()
  const demo = renderDemo()

  return (
    <li className={cn(styles.project, className)}>
      <h3 className={styles.title}>{name.replace(/[.\-/\\\s]/g, " ")}</h3>
      <p className={styles.description}>{description}</p>
      <div className={styles.bottom}>
        <a
          href={html_url}
          className={cn(styles.link, "inline-link")}
          target="_blank"
          rel="noreferrer"
        >
          Go to repository
        </a>
        {demo}
        {tags}
      </div>
    </li>
  )
}
