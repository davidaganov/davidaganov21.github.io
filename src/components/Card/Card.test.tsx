import "@testing-library/jest-dom"
import { render, screen } from "@testing-library/react"
import { Card } from "./Card"

const data = {
  id: "1",
  name: "Test",
  description: "Description project",
  html_url: "repo",
  topics: ["first", "second"],
  homepage: "demo"
}

describe("Card component", () => {
  it("Card renders", () => {
    render(<Card card={data} />)

    expect(screen.getByText(/test/i)).toBeInTheDocument()
  })

  it("There is a link to live demo", () => {
    render(<Card card={data} />)

    expect(screen.getAllByRole("link")).toHaveLength(2)
  })

  it("No link to live demo", () => {
    render(<Card card={{ ...data, homepage: "" }} />)

    expect(screen.getAllByRole("link")).toHaveLength(1)
  })

  it("Card snapshot", () => {
    const card = render(<Card card={data} />)

    expect(card).toMatchSnapshot()
  })
})
