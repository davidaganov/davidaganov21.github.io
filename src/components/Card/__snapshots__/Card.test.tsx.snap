// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component Card snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <li
        class="project"
      >
        <h3
          class="title"
        >
          Test
        </h3>
        <p
          class="description"
        >
          Description project
        </p>
        <div
          class="bottom"
        >
          <a
            aria-label="projects.go_repo Test"
            class="link inline-link"
            href="repo"
            rel="noreferrer"
            target="_blank"
          >
            projects.go_repo
          </a>
          <a
            aria-label="projects.go_demo Test"
            class="link link--live inline-link"
            href="demo"
            rel="noreferrer"
            target="_blank"
          >
            projects.go_demo
          </a>
          <ul
            class="tags"
          >
            <li>
              first
            </li>
            <li>
              second
            </li>
          </ul>
        </div>
      </li>
    </div>
  </body>,
  "container": <div>
    <li
      class="project"
    >
      <h3
        class="title"
      >
        Test
      </h3>
      <p
        class="description"
      >
        Description project
      </p>
      <div
        class="bottom"
      >
        <a
          aria-label="projects.go_repo Test"
          class="link inline-link"
          href="repo"
          rel="noreferrer"
          target="_blank"
        >
          projects.go_repo
        </a>
        <a
          aria-label="projects.go_demo Test"
          class="link link--live inline-link"
          href="demo"
          rel="noreferrer"
          target="_blank"
        >
          projects.go_demo
        </a>
        <ul
          class="tags"
        >
          <li>
            first
          </li>
          <li>
            second
          </li>
        </ul>
      </div>
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
